### PR TITLE
feat(plugin): support  prometheus scrape metrics input  plugins for F…

### LINKED ISSUE #348: Add the PrometheusRemoteWrite output plugin
### Is your feature request related to a problem? Please describe.

Fluent operator already supports node exporter metrics input plugin and opentelemetry output plugin in https://github.com/fluent/fluent-operator/pull/345 

The popular Prometheus remote write plugin should also be supported which enables users to send node exporter metrics to remote Prometheus long-term storages such as Thanos, Cortex, Mimir, or cloud services like grafana cloud

https://docs.fluentbit.io/manual/pipeline/outputs/prometheus-remote-write

### Describe the solution you'd like

And plugin called `PrometheusRemoteWrite` for fluentbit ClusterOutput

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tonyzaizai**
…luent Bit (#355)

              support  prometheus remote write output  plugins for Fluent Bit (#348)

### Which issue(s) this PR fixes:

Fixes #348 and #355 

### Does this PR introduced a user-facing change?

```release-note

```

### Additional documentation, usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by benjaminhuo** [comment]
@tonyzaizai you'll need to sign your commit like below:
```shell
git commit -s --amend
git push -f
```

**Comment by wenchajun** [comment]
DCO failed, please commit with "-s" to add a signature.

Ref:
```
git config --global user.name <name>
git config --global user.email <email>
git commit -s -m "xxx"
```

**Comment by benjaminhuo** [comment]
@tonyzaizai We need to add hostnetwork support in this PR for a fluentbit pod to scrape metrics on the host :
image

**Comment by benjaminhuo** [comment]
> @tonyzaizai We need to add hostnetwork support in this PR for a fluentbit pod to scrape metrics on the host, here 127.0.0.1 is the host's network, not the pod itself: image

@tonyzaizai See also: 
https://v1-22.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#podspec-v1-core
https://github.com/fluent/fluent-operator/blob/master/apis/fluentbit/v1alpha2/fluentbit_types.go#L30
https://github.com/fluent/fluent-operator/blob/master/pkg/operator/daemonset.go#L31

**Comment by wanjunlei** [comment]
> > @tonyzaizai We need to add hostnetwork support in this PR for a fluentbit pod to scrape metrics on the host, here 127.0.0.1 is the host's network, not the pod itself: image
> 
> @tonyzaizai See also: https://v1-22.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#podspec-v1-core https://github.com/fluent/fluent-operator/blob/master/apis/fluentbit/v1alpha2/fluentbit_types.go#L30 https://github.com/fluent/fluent-operator/blob/master/pkg/operator/daemonset.go#L31

Is it possible to access the metrics server on the host using node ip? The hostnetwork mode is restricted in many scenarios.

**Comment by benjaminhuo** [comment]
> Is it possible to access the metrics server on the host using node ip? The hostnetwork mode is restricted in many scenarios.

We add this for KubeEdge edge node container metrics collection only where metrics-server is not available in the edge node.

Edged (KubeEdge's kubelet) expose cadvisor metrics in `<edgenode>:10350/metrics/cadvisor` path

We'll deploy a separate daemonset to collect edge metrics which is different with the cloud daemonset

**Comment by wanjunlei** [comment]
FluentBit support using environment variable in the config like this.

```
[INPUT]
    name prometheus_scrape
    host ${HOST_IP}
    port 8201
    tag vault 
    metrics_path /v1/sys/metrics?format=prometheus 
    scrape_interval 10s
```

We can set the host IP to the pod like this


```
spec:
  containers:
  - env:
    - name: HOST_IP
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: status.hostIP
    image: kubesphere/fluent-bit:v1.8.11
```

The `host` in `prometheus_scrape` can be nil, if the host is nil, or localhost, or 127.0.0.1, fluent operator will replace it with `${HOST_IP}`.

Using this way the `prometheus_scrape` input plugin can apply to all scenarios.

**Comment by benjaminhuo** [comment]
> FluentBit support using environment variable in the config like this.
> 
> ```
> [INPUT]
>     name prometheus_scrape
>     host ${HOST_IP}
>     port 8201
>     tag vault 
>     metrics_path /v1/sys/metrics?format=prometheus 
>     scrape_interval 10s
> ```
> 
> We can set the host IP to the pod like this
> 
> ```
> spec:
>   containers:
>   - env:
>     - name: HOST_IP
>       valueFrom:
>         fieldRef:
>           apiVersion: v1
>           fieldPath: status.hostIP
>     image: kubesphere/fluent-bit:v1.8.11
> ```
> 
> The `host` in `prometheus_scrape` can be nil, if the host is nil, or localhost, or 127.0.0.1, fluent operator will replace it with `${HOST_IP}`.
> 
> Using this way the `prometheus_scrape` input plugin can apply to all scenarios.

Sounds good.
@tonyzaizai, would you help to adjust the host IP logic based on @wanjunlei 's description? We needn't add hostnetwork field this way

**Comment by tonyzaizai** [comment]
ok

> > FluentBit support using environment variable in the config like this.
> > ```
> > [INPUT]
> >     name prometheus_scrape
> >     host ${HOST_IP}
> >     port 8201
> >     tag vault 
> >     metrics_path /v1/sys/metrics?format=prometheus 
> >     scrape_interval 10s
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > We can set the host IP to the pod like this
> > ```
> > spec:
> >   containers:
> >   - env:
> >     - name: HOST_IP
> >       valueFrom:
> >         fieldRef:
> >           apiVersion: v1
> >           fieldPath: status.hostIP
> >     image: kubesphere/fluent-bit:v1.8.11
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The `host` in `prometheus_scrape` can be nil, if the host is nil, or localhost, or 127.0.0.1, fluent operator will replace it with `${HOST_IP}`.
> > Using this way the `prometheus_scrape` input plugin can apply to all scenarios.
> 
> Sounds good. @tonyzaizai, would you help to adjust the host IP logic based on @wanjunlei 's description? We needn't add hostnetwork field this way

ok

File: apis/fluentbit/v1alpha2/plugins/output/prometheus_remote_write_types.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	Uri string `json:"uri,omitempty"`
+	//Add a HTTP header key/value pair. Multiple headers can be set.
+	Headers map[string]string `json:"headers,omitempty"`
+	//Log the response payload within the Fluent Bit log,default: false
+	LogResponsePayload *bool `json:"logResponsePayload,omitempty"`
+	//This allows you to add custom labels to all metrics exposed through the prometheus exporter. You may have multiple of these fields
+	AddLabels []string `json:"addLabels,omitempty"`
+
+	//to do ?
+	Match string `json:"match,omitempty"`
```
Comment (wenchajun) [inline apis/fluentbit/v1alpha2/plugins/output/prometheus_remote_write_types.go:38]: Why do you want to add this field?
It's a repeat of the one here.
https://github.com/fluent/fluent-operator/blob/master/apis/fluentbit/v1alpha2/clusteroutput_types.go#L38

Comment (tonyzaizai) [reply apis/fluentbit/v1alpha2/plugins/output/prometheus_remote_write_types.go:38]: I'll delete it

**Comment by wanjunlei** [comment]
Maybe we should allow users to set `host` to `localhost`, `127.0.0.1` or nil and then replace it with `${HOST_IP}` via fluent operator？

**Comment by tonyzaizai** [comment]
> Maybe we should allow users to set `host` to `localhost`, `127.0.0.1` or nil and then replace it with `${HOST_IP}` via fluent operator？

I  got it

**Comment by benjaminhuo** [comment]
> Maybe we should allow users to set `host` to `localhost`, `127.0.0.1` or nil and then replace it with `${HOST_IP}` via fluent operator？

I'm thinking maybe we can use another work to specify host node, maybe just `host` instead of `localhost` or `127.0.0.1`

**Comment by tonyzaizai** [comment]
> 

ok

**Comment by tonyzaizai** [comment]
> > Maybe we should allow users to set `host` to `localhost`, `127.0.0.1` or nil and then replace it with `${HOST_IP}` via fluent operator？
> 
> I'm thinking maybe we can use another word to specify the host node, maybe simply `host` or `hostnode` instead of `localhost` or `127.0.0.1`
> 
> And the host field cannot be empty or nil, the user must set it explicitly

ok

File: apis/fluentbit/v1alpha2/plugins/input/prometheus_scrape_metrics_type.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	return "prometheus_scrape_metrics"
+}
+
+// Params implement Section() method
+func (p *PrometheusScrapeMetrics) Params(_ plugins.SecretLoader) (*params.KVs, error) {
+	kvs := params.NewKVs()
+	if p.Tag != "" {
+		kvs.Insert("tag", p.Tag)
+	}
+	if p.Host == "" || p.Host == "localhost" || p.Host == "127.0.0.1" {
```
Comment (benjaminhuo) [inline apis/fluentbit/v1alpha2/plugins/input/prometheus_scrape_metrics_type.go:42]:
```suggestion
        host := strings.ToLower(p.Host)
	if host == "" || host == "localhost" || host == "127.0.0.1" || host == "host" {
```

File: apis/fluentbit/v1alpha2/plugins/input/prometheus_scrape_metrics_type.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+		kvs.Insert("host", "${HOST_IP}")
+	} else if strings.ToLower(p.Host) == "host" || strings.ToLower(p.Host) == "hostnode" {
```
Comment (benjaminhuo) [inline apis/fluentbit/v1alpha2/plugins/input/prometheus_scrape_metrics_type.go:44]:
```suggestion
```

File: apis/fluentbit/v1alpha2/plugins/input/prometheus_scrape_metrics_type.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		kvs.Insert("host", "${HOST_IP}")
```
Comment (benjaminhuo) [inline apis/fluentbit/v1alpha2/plugins/input/prometheus_scrape_metrics_type.go:45]:
```suggestion
```

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

**Pull request merged**